flowz/test/artifacts/artifacts_test.py

```diff
@@ -0,0 +1,295 @@
+from collections import OrderedDict
+from concurrent import futures
+
+from nose import tools
+from tornado import gen
+from tornado import testing as tt
+import tornado.concurrent
+
+from flowz.artifacts import (ExtantArtifact, DerivedArtifact, ThreadedDerivedArtifact,
+                             WrappedArtifact, TransformedArtifact, KeyedArtifact,
+                             maybe_artifact)
+from flowz.channels import ChannelDone
+
+
+class ArtifactsTest(tt.AsyncTestCase):
```

Comment:
I think with a bit of adjustment we could get clearer organization and better coverage of the various artifact types.

First, there are a variety of behaviors that all artifacts have.  We ought to exercise this behavior across each them, ideally.  For instance:
- Every artifact has ensure()
- Every artifact has exists()
- Every artifact has get()
- Every artifact has various "as channel" representations.

The classes vary by the underlying behaviors of those, but there's a small handful of high-level cases and all the implementations belong in one of the two or three such cases.

So I think you would want an artifact test that exercises the common behaviors, given a concrete implementation that knows how to build the desired form of artifact.

Then you would have per-implementation subclasses of that test class which could be refined per implementation.

The current thing tests many of the behaviors but doesn't ensure that each artifact implementation gets the full range of expected behaviors.  At least, that's my current reading of it.  I think if you organized it a little differently as described you'd end up with something that was easier to extend in the future, too.

Does that sound reasonable?